### PR TITLE
Add `clone` method for beams and elements

### PR DESCRIPTION
**Pull request by Hespe**
## Description

There is currently no reliable way to obtain copies of `Beam` or `Element`. This has lead to `copy.deepcopy` being used in some places which is troublesome in some instances (see the issues linked below).

## Motivation and Context

- [x] I have raised an issue to propose this changs
  - #225
  - #285

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [x] I have reformatted the code and checked that formatting passes (**required**).
- [x] I have have fixed all issues found by `flake8` (**required**).
- [ ] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [x] I have checked that the documentation builds (**required**).

Note: We are using a maximum length of 88 characters per line.

### CODE REVIEW

**Comment by Hespe** [comment]
Would be wise to merge this PR after #284 such that `Beam.clone()` can be marked abstract.

**Comment by Hespe** [comment]
The second test failure should be fixed once #288 is merged.

**Comment by Hespe** [comment]
The second test failure should be fixed once #288 is merged.

**Comment by Hespe** [comment]
@jank324 There are some lines in the plotting tests that change context beyond their own scope:

https://github.com/desy-ml/cheetah/blob/01dbf8e26d69039f0c9135c672f88daf649ba962/tests/test_plotting.py#L16-L20

Not sure if we should adjust that somehow or if its fine in this case.

**Comment by Hespe** [comment]
> Looks like the `TransverseDeflectingCavity` is broken if its used within a `Segment` and not on its own. The problem seems to be that `Segment.track()` calls `super.track()`, that is `Element.track()`, if the `Segment` is skippable. However, `Element.track()` falls back to the transfer map, which is not implemented for the cavity.

Raised this issue in #290

**Comment by jank324** [comment]
#288 is merged now.

**Comment by Hespe** [comment]
The plotting related failure is now fixed. This leaves only #290 as a blocker for this PR.

File: cheetah/accelerator/rbend.py
Comment (jank324) [inline cheetah/accelerator/rbend.py:61]: @cr-xu we are struggling with this PR a bit and I think you might have useful input.

In `RBend` there might be an issue with `e1` and `e2`. If I instantiate an `RBend` and then access `my_rbend.e1`, the `e1` I get is different from the one I set in the instantiation, by being in `Dipole` space rather than `RBend` space. That's an issue, right?

@Hespe and I were brainstorming solutions and one idea was to define two different `e1` and `e2` ... something like `dipole_e1` and `rbend_e1`, where `rbend_e1` would just be a computed property of the `RBend` class that computes based on `dipole_e1`.

What are your thoughts on this? Are you aware of any conventions already present in other codes or in general notation?

**Comment by Hespe** [comment]
I think the generic `clone()` implementation will not work for the `CustomTransferMap`. The problem is that the important property is `_transfer_map` but the constructor parameter does not have the underscore in front.

**Comment by jank324** [comment]
> I think the generic `clone()` implementation will not work for the `CustomTransferMap`. The problem is that the important property is `_transfer_map` but the constructor parameter does not have the underscore in front.

Ugh ... yeah. I guess we either add a custom implementation for this special case, or rename the value of the transfer map to distinguish it from the function that every element has.

**Comment by Hespe** [comment]
One more thing I've noticed: `name` does not seem to be part of `defining_features` for any of the elements. Cloning would therefore remove all name information at the moment.

File: cheetah/accelerator/bpm.py
Comment (jank324) [inline cheetah/accelerator/bpm.py:64]: For elements where `is_active` can be set manually, shouldn't it be part of `defining_properties`, i.e. wouldn't we want it to also be included in when lattices are saved as well?

**Comment by jank324** [comment]
> One more thing I've noticed: `name` does not seem to be part of `defining_features` for any of the elements. Cloning would therefore remove all name information at the moment.

It probably makes sense for that to be a defining feature of the `Element` base class already, right?

File: cheetah/accelerator/bpm.py
Comment (Hespe) [reply cheetah/accelerator/bpm.py:64]: I had thought that I applied that change a few minutes ago. Maybe something went wrong with that

Comment (Hespe) [reply cheetah/accelerator/bpm.py:64]: Ah, I misunderstood. Is your suggestion to apply the same change to other elements with a manually set `is_active` property?

**Comment by jank324** [comment]
@cr-xu Does this PR now also fix #298?

**Comment by jank324** [comment]
@Hespe I can't officially request a review, so here is an unofficial request.

**Comment by cr-xu** [comment]
> @cr-xu Does this PR now also fix #298?

Feature / method-wise yes. 
I think we should make it clear in the documentation how these values are handled, but this can be done afterwards.

**Review by jank324 — APPROVED** [review]

File: cheetah/particles/particle_beam.py
Comment (Hespe) [inline cheetah/particles/particle_beam.py:1082]: I suppose the survival probability is missing from this list? We merged the other PR here after I had written the `clone` method.

File: tests/test_clone.py
Comment (Hespe) [inline tests/test_clone.py:35]: Initializing the beam with the standard arguments might be a bad idea now that I think about it. If a property is absent from `clone`, it will be initialized by its default and we will not catch that problem here.

**Comment by Hespe** [comment]
Looks good to me except for these two comments about potentially missing features in the `clone` implementation.

File: tests/test_clone.py
Comment (jank324) [reply tests/test_clone.py:35]: Hmm ... maybe the test in general is a bit inconsistent. The name and docstring suggest that it only tests buffer location (where the defaults wouldn't be an issue I think), but it also checks the values

File: cheetah/accelerator/dipole.py
Comment (cr-xu) [inline cheetah/accelerator/dipole.py:1]: The docstring for dipole also needs to be changed e1, e2 -> dipole_e1, dipole_e2

File: cheetah/accelerator/rbend.py
Comment (cr-xu) [inline cheetah/accelerator/rbend.py:19]:
```suggestion
    :param rbend_e1: The angle of inclination of the entrance face in rad.
    :param rbend_e2: The angle of inclination of the exit face in rad.
```

**Comment by cr-xu** [comment]
Apart from that, it's good to go from my side.

**Pull request merged**
